### PR TITLE
[Ready] add stripe integration

### PR DESCRIPTION
**Pull request by dkhgh**
- [x] POC
- [x] write tests
- [x] refactor
- [x] don't create a new customer ALL THE TIME

### CODE REVIEW

File: .sample.env [hunk 1]
```diff
@@ -12,3 +12,6 @@ SMTP_USERNAME=username
 WEB_CONCURRENCY=1
 YAMMER_KEY="client id"
 YAMMER_SECRET="client secret"
+STRIPE_SECRET_KEY="secret key"
+STRIPE_PUBLISHABLE_KEY="publishable key"
+PRICE_PER_USER=2
```
Comment (fatboypunk) [inline .sample.env]: end of line

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -4,6 +4,10 @@ class ApplicationController < ActionController::Base
   protect_from_forgery with: :exception
   helper_method :current_user
 
+  def authorize
+    redirect_to root_path, alert: "Not authorized" if current_user.nil?
```
Comment (fatboypunk) [inline app/controllers/application_controller.rb]: you don't need `.nil?`

File: app/controllers/subscriptions_controller.rb [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+      source: params[:stripeToken]
+    ) 
+
+    redirect_to subscription_path
+  end
+
+  def destroy
+    customer = @stripe_customer
+    card_id = customer.sources.data[0].id 
+    customer.sources.retrieve(card_id).delete()
```
Comment (fatboypunk) [inline app/controllers/subscriptions_controller.rb]: You're using this code in `update` as well as in `destroy`

File: app/models/subscription.rb [hunk 4]
```diff
@@ -0,0 +1,7 @@
+class Subscription < ActiveRecord::Base
+  belongs_to :user
+
+  def cost_per_month
+    cost_per_month = User.count * ENV['PRICE_PER_USER'].to_i
```
Comment (fatboypunk) [inline app/models/subscription.rb]: " :dog:

File: app/views/layouts/application.html.erb [hunk 5]
```diff
@@ -14,8 +14,9 @@
   <%= csrf_meta_tags %>
 </head>
 <body class="<%= body_class %>">
-  <%= render "flashes" -%>
-  <%= yield %>
+  <%= render "flashes" %>
   <%= render "javascript" %>
```
Comment (fatboypunk) [inline app/views/layouts/application.html.erb]: why not add javascript to the head?

File: app/views/static_pages/home.html.erb [hunk 6]
```diff
@@ -1,9 +1,18 @@
...
+
+<nav>
+  <ul>
+    <li><%= link_to "Subscription", subscription_path %></li>
+    <% if current_user %>
+      <li><%= link_to "Log out", signout_path, class: "button" %></li>
+    <% else %>
+      <li><%= link_to "Register", register_path %></li>
+    <% end %>
+  </ul>
```
Comment (fatboypunk) [inline app/views/static_pages/home.html.erb:14]: I think you want to partials for this view: Logged in and logged out

File: app/controllers/application_controller.rb (hunk 2 above)
Comment (dkhgh) [inline app/controllers/application_controller.rb]: and then put `unless current_user`?

File: app/controllers/subscriptions_controller.rb (hunk 3 above)
Comment (dkhgh) [inline app/controllers/subscriptions_controller.rb]: so we should make a `delete_stripe_card` method or something like that? will do, after making the tests

File: app/controllers/application_controller.rb (hunk 2 above)
Comment (fatboypunk) [inline app/controllers/application_controller.rb]: :+1:

File: app/views/layouts/application.html.erb (hunk 5 above)
Comment (dkhgh) [inline app/views/layouts/application.html.erb]: was a quick fix, will do

File: config/initializers/stripe.rb [hunk 7]
```diff
@@ -0,0 +1,7 @@
+Rails.configuration.stripe = {
+  :publishable_key => ENV['STRIPE_PUBLISHABLE_KEY'],
+  :secret_key      => ENV['STRIPE_SECRET_KEY']
```
Comment (fatboypunk) [inline config/initializers/stripe.rb]: old syntax

File: bin/setup [hunk 8]
```diff
@@ -11,7 +11,7 @@ gem install bundler --conservative
 bundle check || bundle install
 
 # Move .sample.env to .env
-mv -n .sample.env .env
+cp -n .sample.env .env
```
Comment (fatboypunk) [inline bin/setup:14]: :flushed:

File: app/controllers/subscriptions_controller.rb (hunk 3 above)
Comment (fatboypunk) [inline app/controllers/subscriptions_controller.rb]: Yes :)

File: app/views/static_pages/home.html.erb (hunk 6 above)
Comment (dkhgh) [inline app/views/static_pages/home.html.erb:14]: yeah nice, will do in another PR though

Comment (fatboypunk) [inline app/views/static_pages/home.html.erb:14]: :+1:

**Comment by dkhgh** [comment]
So let's
- [x] put the stubbed responses in separate .json files
- [x] load the api key for the stubs from ENV
- [x] test all the stripe things we do
- [x] refactor the `subscriptions_controller`

**Comment by fatboypunk** [comment]
> put the stubbed responses in separate .json files
>  load the api key for the stubs from ENV
>  test all the stripe things we do
>  refactor the subscriptions_controller
- [x] Move the stub_requests to a support file

**Comment by dkhgh** [comment]
Time to refactor :candy:

File: spec/spec_helper.rb [hunk 9]
```diff
@@ -32,6 +32,7 @@ module Features
 
   config.order = :random
   config.include Features, type: :feature
+  config.include StripeHelper
```
Comment (fatboypunk) [inline spec/spec_helper.rb:35]: this shouldn't be needed as long as it's in `support/**/**.rb`

Comment (dkhgh) [inline spec/spec_helper.rb:35]: I think it was necessary.. didn't work before

File: app/controllers/subscriptions_controller.rb [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+    Subscription.transaction do
+      customer = Stripe::Customer.create(
+        source: params[:stripeToken],
+        plan: "ca-basic",
+        quantity: User.count,
+        email: current_user.email
+      )
+      resource.update(
+      stripe_id: customer.id,
+      subscription_id: customer.subscriptions.data[0].id
```
Comment (fatboypunk) [inline app/controllers/subscriptions_controller.rb]: indentation :dog:

File: app/models/subscription.rb [hunk 11]
```diff
@@ -0,0 +1,7 @@
+class Subscription < ActiveRecord::Base
+  belongs_to :user
+
+  def cost_per_month
+    cost_per_month = User.count * ENV["PRICE_PER_USER"].to_i
+  end
+end
```
Comment (fatboypunk) [inline app/models/subscription.rb]: end of line :poodle:

File: app/views/subscriptions/new.html.erb [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+    } else {
+      // response contains id and card, which contains additional card details
+      var token = response.id;
+      // Insert the token into the form so it gets submitted to the server
+      $form.append($('<input type="hidden" name="stripeToken" />').val(token));
+      // and submit
+      $form.get(0).submit();
+    }
+  };
+</script>
```
Comment (fatboypunk) [inline app/views/subscriptions/new.html.erb:58]: I think this form is now the same as the edit form except from the header

**Comment by dkhgh** [comment]
`fixtures/vcr_cassettes` still needs to go.

File: app/views/subscriptions/new.html.erb (hunk 12 above)
Comment (dkhgh) [inline app/views/subscriptions/new.html.erb:58]: And the method, `:post` vs `:put`.
Also, I really don't like reusing form partials.

File: app/models/subscription.rb (hunk 11 above)
Comment (dkhgh) [inline app/models/subscription.rb]: fixed, but doesn't show up here.

**Comment by dkhgh** [comment]
`fixtures/vcr_cassettes` is gone.

File: app/views/subscriptions/new.html.erb (hunk 12 above)
Comment (fatboypunk) [inline app/views/subscriptions/new.html.erb:58]: also an end of line here, think we could make this a form_for with a subscription model.
And i do like forms in a DRY way, only when it gets to hard to keep them in the same place it's good to split them up

File: lib/tasks/update_subscription_quantity.rake [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+    subscription_id = Subscription.last.subscription_id
+    customer = Stripe::Customer.retrieve(stripe_id)
+    subscription = customer.subscriptions.retrieve(subscription_id)
+    subscription.quantity = User.count
+    subscription.save
+    puts "updated the subscription quantity to #{User.count}"
+  else
+    puts "there's no subscription to update"
+  end
+end
```
Comment (fatboypunk) [inline lib/tasks/update_subscription_quantity.rake]: end of line :dog2:

File: spec/features/user_logs_in_with_oauth_spec.rb [hunk 14]
```diff
@@ -8,6 +8,6 @@
 
     click_link "Register with Yammer"
 
-    expect(page).to have_content("Welcome: foo")
+    expect(page).to have_content("Welcome: Han Solo")
```
Comment (fatboypunk) [inline spec/features/user_logs_in_with_oauth_spec.rb:11]: we can pass the name to the `set_omniauth` method so i think it's better to set that name in a variable and give it to omniauth and check here if he logged in

**Comment by fatboypunk** [comment]
I made a few comments, but we could merge this already

File: app/views/subscriptions/new.html.erb (hunk 12 above)
Comment (dkhgh) [inline app/views/subscriptions/new.html.erb:58]: :+1: work for another PR though

**Pull request merged**